File: app/javascript/channels/line_channel.js
import consumer from "./consumer"

document.addEventListener("turbo:load", () => {
  const comments = document.getElementById('comments');
  if (comments === null) {
    return;
  }
  const currentUserId = document.body.dataset.currentUserId;
  const lineId = comments.dataset.lineId;
  const appLine = consumer.subscriptions.create({channel: "LineChannel", line_id: lineId}, {
    received(data) {
      let commentHtml;
      if (data.sender_id.toString() === currentUserId) {
        commentHtml = data.comment;
      } else {
        commentHtml = data.second_comment;
      }
      
      comments.insertAdjacentHTML('beforeend', commentHtml);
    },
  
    speak: function(comment, lineId) {
      return this.perform('speak', {comment: comment, line_id: lineId});
    }
  });

  const input = document.getElementById('line_input');
  const button = document.getElementById('submit_button');

  if(input) {
    input.addEventListener("keypress", function(e) {
      if (e.key === 'Enter') {
        appLine.speak(e.target.value, lineId);
        e.target.value = '';
        e.preventDefault();
      }
    });
  } else {
    console.error("Could not find element with id 'line_input'");
  }

  if(button) {
    button.addEventListener("click", function(e) {
      if(input.value !== '') {
        appLine.speak(input.value, lineId);
        input.value = '';
      }
      e.preventDefault();
    });
  } else {
    console.error("Could not find element with id 'submit_button'");
  }
});

File: app/channels/line_channel.rb
class LineChannel < ApplicationCable::Channel
  def subscribed
    line = Line.find(params[:line_id])
    stream_for line
  end

  def unsubscribed
  end

  def speak(data)
    line = Line.find(data['line_id'])
    body = data['comment'].strip # メッセージの前後の空白を削除
    return unless body.present? # メッセージが空でないことを確認

    # メッセージの作成だけ行い、ブロードキャストは after_commit フックに任せる
    comment = line.comments.create!(body:, user_id: current_user.id)
    sleep(0.1)
  end
end
